.circle/index.py

```diff
@@ -105,27 +123,87 @@ def build_index(path_glob, output_path):
     print('Index data written to "%s".' % (output_path))
 
 
+def get_available_versions():
+    """
+    Retrieve all the available versions for all packs
+
+    NOTE: This function uses Github API.
+    """
+    pages = []
+
+    # use `gh` to handle getting all pages
+    proc = subprocess.Popen(
+        [
+            "gh", "api", "--paginate", "graphql",
+            "-f", "owner=" + EXCHANGE_NAME,
+            "-f", "query=" + GQL_PACK_TAGS_QUERY,
+        ],
+        env={"GH_TOKEN": GITHUB_PASSWORD},
+        stdout=subprocess.PIPE,
+    )
+    # This should never take more than 5 seconds.
+    # If network is really bad, let it go for 30.
+    try:
+        outs, _ = proc.communicate(timeout=30)
+    except subprocess.TimeoutExpired:
+        proc.kill()
+        outs, _ = proc.communicate()
+    result = outs.decode().strip()
+
+    # https://stackoverflow.com/a/43807246/1134951
+    decoder = json.JSONDecoder()
+    pos = 0
+    len_result = len(result)
+    while pos < len_result:
+        j, json_len = decoder.raw_decode(result, idx=pos)
+        pos += json_len
+        pages.append(j)
+
+    # PREFIX-<pack name>
+    prefix_len = len(EXCHANGE_PREFIX) + 1
+
+    for page in pages:
+        repos = page["data"]["repositoryOwner"]["repositories"]["nodes"]
+        packs = [
+            {
+                "reponame": repo["name"],
+                "tags": [tag["name"] for tag in repo["refs"]["nodes"]]
+            } for repo in repos if repo["name"].startswith(EXCHANGE_PREFIX)
+        ]
+        for pack in packs:
+            pack_name = pack["reponame"][prefix_len:]
+            if pack_name not in PACK_VERSIONS:
+                PACK_VERSIONS[pack_name] = []
+            PACK_VERSIONS[pack_name].extend(
+                tag.replace("v", "") for tag in pack["tags"]
+                if tag.startswith("v")  # only version tags
+            )
```

Comment:
This is a great place to use `dict.setdefault`:

```suggestion
            PACK_VERSIONS.setdefault(pack_name, []).extend(
                tag.replace("v", "") for tag in pack["tags"]
                if tag.startswith("v")  # only version tags
            )
```

Example:

```python
>>> d = {'a': [1], 'b': [2]}
>>> d.setdefault('a', []).extend([3, 4, 5])
>>> d.setdefault('c', []).extend([6, 7, 8])
>>> d
{'a': [1, 3, 4, 5], 'c': [6, 7, 8], 'b': [2]}
```